### ISSUE TITLE
DateTime unsafely uses private members of arguments.

### ISSUE THREAD

**Issue by sanderAtMaxdoro** [dart-lang/sdk#34962]
Thank you for taking the time to file an issue!

In order to route, prioritize, and act on this, please include:

* Dart SDK Version (`dart --version`)
Flutter 0.10.2-pre.47 • channel master • https://github.com/flutter/flutter.git
Framework • revision 9dfc0f3aaa (5 days ago) • 2018-10-24 07:22:08 +0100
Engine • revision 2586e94122
Tools • Dart 2.1.0-dev.7.1.flutter-45f9462398

* Whether you are using Windows, MacOSX, or Linux (if applicable)
MacOSX

* Whether you are using Chrome, Safari, Firefox, Edge (if applicable)
NA

Missing some or all of the above might make the issue take longer or be
impossible to act on.

----

Is it really an issue? For general questions consider starting with Stack
Overflow:
https://stackoverflow.com/questions/tagged/dart

Also consider our Gitter channel for light-weight/quick discussions:
https://gitter.im/dart-lang/sdk

-----
When you let a class implement another class using

class MyClass implements YourClass

the

final <Type> _propertyName;

vars are part of the public class interface. In particular, MyClass is expected to implement  _propertyName by the implementation of YourClass, even though there is no proper way to implement that when you look at the public interface. You can of course see it in the code if you have the code and add it, but that is fraud (in the Stroustrup sense).

Try this by implementing your own MyDateTime as an implementation of DateTime. MyDateTime is expected to implement final int _value, and a NoSuchMethod is thrown if it is not found by DateTime.

**Comment by lrhn** [comment]
If I understand this correctly, the issue is that `DateTime` assumes that every implementation of the interface also has a `_value` getter. That's obviously not true if the interface is implemented in a different library. The code that makes this assumption should be written to work in terms of the public interface of the class.

Dart does allow implementing interfaces that have private and inaccessible members. That is generally unsafe, but only to the library that declared this public interface with private members, the rest of the world can't see the members anyway.

**Comment by lrhn** [comment]
The proper behavior is to call `Bar.noSuchMethod` with `Invocation.getter(#_f)` where `#_f` is the symbol from `foo.dart`. (I'll write that symbol as `#foo::_x` for now, there is obviously no way to write it directly in `bar`)

More precisely: The class `Bar` does not implement `int get _x` or `set _x(int _)`, and since the members are inaccessible, that is accepted and noSuchMethod-forwarders are inserted:
```dart
  int get foo:_x => this.noSuchMethod(Invocation.getter(#foo::_x);
  void set foo::_x(int $) => this.noSuchmethod(Invocation.setter(#foo::_x, $);
```
These are the same type-checking `noSuchMethod` forwarders inserted for non-abstract classes that override `noSuchMethod`, but we introduce them for inaccessible members even when `noSuchMethod` has not been overridden. This ensures that any non-abstract class implements its entire interface, even the parts it doesn't know about, it just has to do that by throwing. It's very close to the Dart 1 behavior, but still implementable using a v-table.

The issure filed here is not about that, it's just that `DateTime` accesses such private members on objects that it doesn't know to have them implemented. I'm fixing that.


Now, one case I'm not sure we handle correctly (@eernstg):
```dart
library lib1;
class C {
  int _foo(int x) => x;
}
abstract class D {
  int _foo([int x]);
}
```
and
```dart
library lib2;
import "lib1.dart";
class E extends C implements D {}
```
Will class `E` get a noSuchMethod forwarder overriding the concrete `C._foo` (which we otherwise never do), or will we reject the program for reasons inaccessible to the lib2 programmer?

**Comment by eernstg** [comment]
Lasse and I just discussed this, and we agreed that the only meaningful approach is to generate the noSuchMethod forwarder, even though it will override a user-written concrete declaration.

When we otherwise refuse to generate an nSM forwarder _m<sub>1</sub>_ that overrides a user-written concrete declaration _m_, it is because that forwarder _should_ be written by a human being&mdash;_m<sub>1</sub>_ will always be able to forward some of the calls that it might receive to _m_, and it's a matter of human judgment whether/when to do that. With the inaccessible methods the developer just can't write the forwarder anyway, so it makes no sense to require human help.

I created [this CL](https://dart-review.googlesource.com/c/sdk/+/82042) adding the missing word 'accessible' to dartLangSpec.tex, such that this case is covered explicitly.

That said, we should also have a lint for detecting that "I'm now writing a class `D` that's supposed to masquerade as `library2.C` based on `implements library2.C`, but there is no way I can expect to pass a `D` to code in `library2` and have it work correctly, because `library2.C` has private methods". It might be fine to allow classes like `D` in scenarios where the usage of a supposed `library2.C` object is simple and well-understood (say, in a test), but we should also help developers avoid creating them in the first place if they intend to create a full-fledged replacement for `library2.C`.

**Comment by eernstg** [comment]
Said CL was landed as 117c23fddbfa66068a56fef171188fd0f41745c2, which means for Lasse's class `E` that the nSM forwarder is generated and there is no error. Removing 'area-language'.

I suppose we should change this to an 'area-meta' issue, and create tool specific issues as subtasks.

**Comment by lrhn** [comment]
Title changed. CL to fix the issue: https://dart-review.googlesource.com/c/sdk/+/81828
